airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/per_partition_cursor.py

```diff
@@ -0,0 +1,260 @@
+#
+# Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+#
+
+from typing import Any, Iterable, Mapping, Optional
+
+from airbyte_cdk.models import SyncMode
+from airbyte_cdk.sources.declarative.stream_slicers.stream_slicer import StreamSlicer
+from airbyte_cdk.sources.declarative.types import Record, StreamSlice, StreamState
+
+
+class Hashabledict(dict):
```

Comment:
Those are good points! I gave it a shot with the flattened tuples and it addresses the flaws you were mentioning. Let me know what you think

Reply:
Hmm, agree this needs some thought. One question that I'm getting hung up on is whether this is any different than what we already deal with when we serialize/deserialize the state message as json, since json doesn't have tuples. Do you know whether we're already dealing with this kind of situation?

Reply:
We could probably serialize this as JSON given that we iron out some details. One thing I see if the order of fields. In JSON, the order of fields is not relevant regarding equality. However, once parse as a string, this is relevant. The solution with tuples solved this by ordering keys. I'm wondering if there would be other things to consider if we were to use JSON

Reply:
Does ensuring that tuples aren't part of state feel tractable to you? Is it possible that we already do that, since the state has to be json serializable and tuples are not json serializable?

One thing I'm not totally sure of is whether we're talking about rare and hypothetical scenarios. Do you have an opinion on that given your experience in this area? I feel like if the list/tuple situation is sufficiently rare, restricting it could be okay (especially if there are workarounds).